kube-runtime/src/controller/runner.rs

```diff
@@ -170,8 +220,108 @@ mod tests {
         // a timeout here *should* mean that the background task isn't getting awoken properly
         // when the new message is ready.
         assert_eq!(
-            timeout(Duration::from_secs(1), result_rx).await.unwrap().unwrap(),
+            timeout(Duration::from_secs(1), result_rx)
+                .await
+                .unwrap()
+                .unwrap()
+                .transpose()
+                .unwrap(),
             Some(8)
         );
     }
+
+    #[tokio::test]
+    async fn runner_should_wait_for_readiness() {
+        let is_ready = Mutex::new(false);
+        let (delayed_init, ready) = DelayedInit::<()>::new();
+        let mut runner = Box::pin(
+            Runner::new(
+                scheduler(
+                    stream::iter([ScheduleRequest {
+                        message: 1u8,
+                        run_at: Instant::now(),
+                    }])
+                    .chain(stream::pending()),
+                ),
+                |msg| {
+                    assert!(*is_ready.lock().unwrap());
+                    future::ready(*msg)
+                },
+            )
+            .delay_tasks_until(ready.get()),
+        );
+        assert!(poll!(runner.next()).is_pending());
+        *is_ready.lock().unwrap() = true;
+        delayed_init.init(());
+        assert_eq!(runner.next().await.transpose().unwrap(), Some(1));
+    }
+
+    #[tokio::test]
+    async fn runner_should_dedupe_while_waiting_for_readiness() {
+        let is_ready = Mutex::new(false);
+        let (delayed_init, ready) = DelayedInit::<()>::new();
+        let mut runner = Box::pin(
+            Runner::new(
+                scheduler(
+                    stream::iter([
+                        ScheduleRequest {
+                            message: 1u8,
+                            run_at: Instant::now(),
+                        },
+                        ScheduleRequest {
+                            message: 2u8,
+                            run_at: Instant::now(),
+                        },
+                        ScheduleRequest {
+                            message: 1u8,
+                            run_at: Instant::now(),
+                        },
+                    ])
+                    .chain(stream::pending()),
+                ),
+                |msg| {
+                    assert!(*is_ready.lock().unwrap());
+                    future::ready(*msg)
+                },
+            )
+            .delay_tasks_until(ready.get()),
+        );
+        assert!(poll!(runner.next()).is_pending());
+        *is_ready.lock().unwrap() = true;
+        delayed_init.init(());
+        assert_eq!(
+            runner.as_mut().take(2).try_collect::<HashSet<_>>().await.unwrap(),
+            HashSet::from([1, 2])
```

Comment:
Is this actually testing that deduplication works? If you are asking for only 2 elements (and into a `HashSet`) wouldn't you get those 2 elements regardless?

Reply:
... yeah, this wasn't very clear. We test on the line below that there's nothing remaining in the runner, but the `HashSet` test would fail with a very confusing error if the condition doesn't hold...